src/main/flight/pos_hold.c

```diff
@@ -0,0 +1,112 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Betaflight. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#ifdef USE_POS_HOLD_MODE
+
+#include "math.h"
+#include "build/debug.h"
+#include "common/maths.h"
+
+#include "config/config.h"
+#include "fc/runtime_config.h"
+#include "fc/rc.h"
+#include "flight/failsafe.h"
+#include "flight/position.h"
+#include "flight/autopilot.h"
+#include "rx/rx.h"
+
+#include "pos_hold.h"
+
+posHoldState_t posHold;
+
+void posHoldReset(void)
+{
+    posHold.posHoldIsOK = true;
+    posHold.targetLocation = gpsSol.llh;
+    resetPositionControl(posHold.targetLocation);
+}
+
+void posHoldInit(void)
+{
+    posHold.isPosHoldRequested = false;
+    posHold.deadband = rcControlsConfig()->pos_hold_deadband / 100.0f;
+    posHold.useStickAdjustment = rcControlsConfig()->pos_hold_deadband;
+}
+
+void posHoldProcessTransitions(void)
+{
+    if (FLIGHT_MODE(POS_HOLD_MODE)) {
+        if (!posHold.isPosHoldRequested) {
+            // initialise relevant values each time position hold starts
+            posHoldReset();
+            posHold.isPosHoldRequested = true;
+        }
+    } else {
+        posHold.isPosHoldRequested = false;
+    }
+}
+
+void posHoldUpdateTargetLocation(void)
+{
+    // if failsafe is active, eg landing mode, don't update the original start point 
+    if (!failsafeIsActive()) {
+        // otherwise if sticks are not centered, allow start point to be updated
+        if (posHold.useStickAdjustment) {
+            if ((getRcDeflectionAbs(FD_ROLL) > posHold.deadband) || (getRcDeflectionAbs(FD_PITCH) > posHold.deadband))  {
+                // allow user to fly the quad, in angle mode, when sticks are outside the deadband
+                // while sticks are outside the deadband,
+                // keep updating the home location to the current GPS location each iteration
+                posHoldReset();
+            }
+        }
+    }
+}
+
+void posHoldUpdate(void)
+{
+    posHoldUpdateTargetLocation();
+
+    if (getIsNewDataForPosHold() && posHold.posHoldIsOK) {
+        posHold.posHoldIsOK = positionControl(posHold.deadband);
+    }
+}
+
+void updatePosHoldState(timeUs_t currentTimeUs) {
+    UNUSED(currentTimeUs); 
+
+    // check for enabling Alt Hold, otherwise do as little as possible while inactive
+    posHoldProcessTransitions();
+
+    if (posHold.isPosHoldRequested) {
+        posHoldUpdate();
+    } else {
+        posHoldAngle[AI_PITCH] = 0.0f;
+        posHoldAngle[AI_ROLL] = 0.0f;
+    }
+}
+
+bool showPosHoldWarning(void) {
+    return (posHold.isPosHoldRequested && !posHold.posHoldIsOK);
+}
+
+bool allowPosHoldWithoutMag(void) {
+    return (posHoldConfig()->pos_hold_without_mag);
+}
```

Comment:
The purpose of `allowPosHoldWithoutMag` is to allow a pilot who does not have a Mag to choose to enable Position Hold if they want to.
The check in core.c achieves that with `sensors(SENSOR_MAG) || allowPosHoldWithoutMag()`.
For that check, `sensors(SENSOR_MAG)` will be accessible, and false, if `USE_MAG` is not defined, and it will also be false if the user says they don't want to use their Mag.

With the suggested change, the result is always false if Mag is not used, or not available.  But we need a test where it can be true when the user wants to continue.

Maybe `return sensors(SENSOR_MAG) || allowPosHoldWithoutMag()` would simplify core.c and maintain the intent?

Reply:
My suggestion does just that. It checks whether posHold is supposed to use the mag. If there is no mag, it obviously will return false to say that posHold is not using the mag (because without mag, posHold cannot use the mag, even if it is supposed to use a mag). You need to invert the logic on `pos_hold_without_mag` for my suggestion to make sense.

Reply:
As I read it, it will *only* return true when:
- mag in build, and 
- mag is enabled, and 
- user says, use the mag.
Otherwise it is false.
We need something that returns true when:
- mag in not in build, or 
- mag is not enabled, but 
- user says, don't use the mag.
Then we let them go ahead.
I don't think your function achieves this.

I currently use this only at the point where pos_hold should be enabled, in core.c, when the switch is activated.  There we need the user to allow it to go ahead even if there is no Mag.  The code is easy to understand.

Separately, if the user has permitted us to go ahead without mag, we use compassIsHealthy(), which is positive when mag is in build, sensor activated, and sensor calibrated and returning numbers.  If we have that, we have a mag, and we don't care about GPS heading.  Otherwise if we don't have a mag, and also we don't have a GPS heading, we fail and alert the user, even though we allowed them to try without the mag.

Reply:
This is exactly what my suggestion achieves. Just that the logic is inverted from your point of view. So if you want to check if posHold is **not** using the mag (for whatever reason, whether it is user preference, broken/undetected mag or simply no mag code compiled), you simply write
```c
!posHoldIsUsingMag()
```
using the logical "NOT-operator" `!`. This will return true, if
- mag in not in build, or
- mag is not enabled, but
- user says, it's OK, I don't care about not having a mag

Reply:
I think I understand now that I appreciate that your suggestion is to invert the logic completely.

So that:
- current version:  Say 'Yes' to `pos_hold_without_mag` (default No)
- proposal: Say 'No' to `pos_hold_use_mag` (default Yes)

Either way is OK, I think it doesn't matter much; whatever text the user finds most intuitive.  The label `pos_hold_use_mag` could perhaps be more intuitive as `pos_hold_only_with_mag` or `pos_hold_requires_mag`.

In the IMU code, the Mag always takes precedence, if available, for heading; the check we are discussing is only currently used to block initiation of the flight mode.  Currently there is no warning, the flight_mode just doesn't start, if there is no Mag and it is required.

TO get a warning to the pilot, the check needs to go in the autopilot code to cause position hold to fail.  Or else we make a dedicated warning for 'mag required but not available' somehow.

Hence I think probably we should always allow the flight mode to activate, but fail if no mag and user requires it; then they get the warning.

I've now tested it without Mag, and it works OK.  The block on initiation until the IMU is oriented seems to work.

If experience is that it works well enough without Mag, we can reverse the default behaviour.

Reply:
@KarateBrot - I've revised the logic so we let the flight mode activate (remove the check in core.c) but do this check in positionControl() so that we get the OSD warning.  Otherwise, we don't get the warning, because the flight mode never starts.

Reply:
@KarateBrot : IMO @ctzsnooze version is fine. The option is to explicitly allow poshold even when mag is not available (for any reason). It should default to 'no mag - no poshold' (and 0 is slightly preferable for default). And it should be quite explicit that user is willing to take the risk, at least for some time.